opennebula/resource_opennebula_virtual_machine.go

```diff
@@ -1712,6 +1725,29 @@ func generateVm(d *schema.ResourceData, tplContext *dyn.Vector) (string, error)
 }
 
 func resourceVMCustomizeDiff(diff *schema.ResourceDiff, v interface{}) error {
+
+	onChange := diff.Get("on_disk_change").(string)
+
+	if strings.ToUpper(onChange) == "RECREATE" {
+
+		oldDisk, newDisk := diff.GetChange("disk")
+		oldDiskList := oldDisk.([]interface{})
+		newDiskList := newDisk.([]interface{})
+		toDetach, _ := diffListConfig(newDiskList, oldDiskList,
+			&schema.Resource{
+				Schema: diskFields(),
+			},
+			"image_id",
+			"target",
+			"driver",
+			"volatile_type",
+			"volatile_format")
+
+		if len(toDetach) > 0 {
+			diff.ForceNew(fmt.Sprintf("disk.%d.image_id", toDetach[0]))
```

Comment:
It works on an `image_id` change, not for other attributes changes, and I'm not sure that it will works in some cases with several disks attached.
If we want the behavior to be triggered only on `image_id` change, we should write this instead:
```
		toDetach, toAttach := diffListConfig(newDiskList, oldDiskList,
			&schema.Resource{
				Schema: diskFields(),
			},
			"image_id")
```

Not sure for the key in `diff.ForceNew`.
The `%d` in `disk.%d.image_id` should be the disk position in the list. We want to enforce the `ForceNew` behavior when a given attribute of the disk has changed.
The problem is that with the `diffListConfig` method we lose the disk position in the lists returned.

I ended up with this kind of brute forcing code, I'm not fully satisfied but this seems to work:
```
	if strings.ToUpper(onChange) == "RECREATE" {

		oldDisk, newDisk := diff.GetChange("disk")
		oldDiskList := oldDisk.([]interface{})
		newDiskList := newDisk.([]interface{})
		toDetach, toAttach := diffListConfig(newDiskList, oldDiskList,
			&schema.Resource{
				Schema: diskFields(),
			},
			"image_id",
			"target",
			"driver")

		if len(toDetach) > 0 {
			for i := range oldDiskList {
				diff.ForceNew(fmt.Sprintf("disk.%d.image_id", i))
				diff.ForceNew(fmt.Sprintf("disk.%d.target", i))
				diff.ForceNew(fmt.Sprintf("disk.%d.driver", i))
			}
		}
```
If we are in the case where we want to detach a disk, this iterate over all disks and call ForceNew on some attributes. If nothing has changed, the ForceNew method returns an error, else, this generate a new diff that show which field changed and triggered the VM recreation.

There is probaly some other solutions, maybe simpler, feel free to test this with several disks

Reply:
@treywelsh I went ahead and went with your suggestion, it seems to work just fine. Do you think we can merge this soon?

Reply:
Sorry for this, I made the test just after he reads the PR.

I won't work on the provider for one week, then we just need to focus on 1 or 2 bug fixes to release the 0.4.2 version (this should be quick) then I'll take a look at your PR, test it and it should be good.

This will be part of the 0.5.0 but there is still some work to do to release it (we plan a bunch of things then we'll need to test it a bit more in depth etc.).

I suppose you need this quickly ?

Reply:
@treywelsh It's going to greatly increase the speed of a deployment process we have, but it's all good. It looks like you're releasing on a cadence of every month or so. In the grand scheme of things, i've been working on it with you since November, so another month or so won't hurt!